### PR TITLE
Add a .usage property on LLM object

### PR DESCRIPTION
**Pull request by hexapode**
Add a .usage property on LLM object allowing tracking of token/sec consuption and Amount accross all call for OpenAI and Anthropic.

Also update the LLM module doc to reflect the change

### CODE REVIEW

**Review by marcusschiesser — COMMENTED** [review]
@hexapode Great idea, but how do we keep this in sync with the price changes?

File: packages/core/src/llm/LLM.ts [hunk 1]
```diff
@@ -45,6 +45,8 @@ export interface ChatResponse {
   message: ChatMessage;
   raw?: Record<string, any>;
   delta?: string;
+  metrics?: any;
+  usage?: any;
```
Comment (himself65) [inline packages/core/src/llm/LLM.ts:49]: don't use `any` if possible

Comment (hexapode) [reply packages/core/src/llm/LLM.ts]: Yes, changed it for usage.

**Comment by hexapode** [comment]
> @hexapode Great idea, but how do we keep this in sync with the price changes?

@marcusschiesser: We have to manually check? No better idea for now / no provider expose a pricing API?

**Comment by marcusschiesser** [comment]
> > @hexapode Great idea, but how do we keep this in sync with the price changes?
> 
> @marcusschiesser: We have to manually check? No better idea for now / no provider expose a pricing API?

@hexapode I think for a start we could just track the usage (i.e. tokens and compute seconds) and not the costs. That's still valuable and if needed a user could calculate the costs - e.g. you could add an example computing the costs in `examples/`

**Comment by EmanuelCampos** [comment]
Agreed, we can just return the total tokens used, the price isn't trustable and we can't let users think that's the current real price

bump @hexapode if you have any plans for it, when I have free time can go with it as well

**Comment by marcusschiesser** [comment]
TODOS:

1. remove `cost` from usage (as the price is not stable)
2. add example (same as in docs)

**Comment by himself65** [comment]
> I will try to use `AsyncLocalStorage` on this

I think the ideal API would be

```ts
const usage = new Usage();
const llm = new OpenAI();
usage.track(llm)
// ...
console.log(usage.cost) // in dollar
```

**Comment by marcusschiesser** [comment]
@himself65, your API looks great!

We probably don't need to support reusing `Usage` objects between LLMs, so you can even simplify it to:
```
const llm = new OpenAI();
const openAIUsage = new Usage(llm);
// ...
console.log(usage.tokens) 

{
	const newLLM = new AnotherAI();
        const anotherAIUsage = new Usage(newLLM);
}
```
Btw: I don't think we should start with the costs, see https://github.com/run-llama/LlamaIndexTS/pull/270#issuecomment-1867576025

**Comment by himself65** [comment]
closing as you should implement this in user land

**Pull request closed without merging**